verilog/analysis/checkers/generate_label_prefix_rule_test.cc

```diff
@@ -0,0 +1,154 @@
+#include "common/analysis/syntax_tree_linter_test_utils.h"
+#include "common/text/symbol.h"
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/CST/verilog_treebuilder_utils.h"
+#include "verilog/analysis/verilog_analyzer.h"
+
+namespace verilog {
+namespace analysis {
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+TEST(GenerateLabelPrefixRuleTest, Various) {
+  const std::initializer_list<LintTestCase> kTestCases = {
+      {""},
+      {"module m;\n"
+       "endmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) ",
+       "begin",
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin : g_label"
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin : gen_label"
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin : gen_label"
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end : gen_label_also\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "begin\n"
+       "  initial begin : g_ini\n"
+       "    if (1) begin : gen_if\n"
+       "    end : gen_endif\n"
+       "  end : g_endini\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      // Test incorrect code
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin : ",
+       {SymbolIdentifier, "k_label"},
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) "
+       "begin",
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end : ",
+       {SymbolIdentifier, "genwithoutunderscore"},
+       "\nendgenerate\nendmodule\n"},
+      {"module m;\n"
+       "generate\n"
+       "if (TypeIsPosedge) ",
+       "begin : ",
+       {SymbolIdentifier, "k_label"},
+       "\n"
+       "  always @(posedge clk) foo <= bar;\n"
+       "end : ",
+       {SymbolIdentifier, "genwithoutunderscore"},
+       "\nendgenerate\nendmodule\n"},
+      // Incorrect code with more blocks
+      {"module m;\n"
+       "generate\n"
+       "begin : ",
+       {SymbolIdentifier, "wrong"},
+       "\n"
+       "  initial begin : g_ini\n"
+       "    if (1) begin : gen_if\n"
+       "    end : gen_endif\n"
+       "  end : g_endini\n"
+       "end\n"
+       "endgenerate\nendmodule\n"},
+
+      {"module m;\n"
+       "generate\n"
+       "begin\n",
+       "  initial begin : ",
+       {SymbolIdentifier, "wrong"},
```

Comment:
This is technically a statement, not a generate construct.  I don't think this rule should cover here, so I would expect this to be a negative test.  (Same with closely related tests)